### DISCUSSION TITLE
Test not found on running `forge test --match-test testIfRaffleDeployed`

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#582] [Q&A] [answered]
I was writting a few extra tests for the Raffle smart contract until I encountered that foundry is not able to pickup one of my tests. Sharing the test below. 
This was an integration test for `DeployRaffle.s.sol` under the directory `/test/integration/DeployRaffleTest.t.sol`

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test} from "forge-std/Test.sol";
import {DeployRaffle} from "../../script/DeployRaffle.s.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";
import {Raffle} from "../../src/Raffle.sol";
import {DevOpsTools} from "@foundry-devops/DevOpsTools.sol";

contract DeployRaffleTest is Test {
    Raffle public raffle;
    HelperConfig public helperConfig;
    address PLAYER = makeAddr("player");
    uint256 STARTING_BALANCE = 10 ether;

    function setUp() external {
        vm.prank(PLAYER);
        vm.deal(PLAYER, STARTING_BALANCE);

        DeployRaffle deployRaffle = new DeployRaffle();
        (raffle, helperConfig) = deployRaffle.run();
    }

    function testIfRaffleDeployed() public {
        address lastestDeployedRaffle = DevOpsTools.get_most_recent_deployment(
            "Raffle",
            block.chainid
        );

        assert(address(raffle) == lastestDeployedRaffle);
    }
}
```

The project compiles well on running `forge-build`.
But on running `forge test --match-test testIfRaffleDeployed` I'm getting the following result:
```
No tests match the provided pattern:
        match-test: `testIfRaffleDeployed`

Did you mean `testRaffleRecordsPlayerWhenTheyEnter`?
```
This is really weird because so far I've written quite a lot of tests and have decent experience with them. I'm sure I'm missing something small and basic but just not able to spot it. 
Please enlighten me 😅. Thanks!

**Comment by theirrationalone** [comment]
It's Not `testRaffleDeployed`, it's `testIfRaffleDeployed`.. 

You're missing that Is with capital I.

**Reply by theirrationalone** [reply]
Read this tooo... 


Still Test may lead to be failed... Why??? 

Because that `DevOpsTools.get_most_recent_deployment` takes that latest contract which has been deployed and stored into broadcast folder using that `--broadcast` cmd. 

You should try different approach to test deploy scripts.

**Reply by alfheimrShiven** [reply]
@theirrationalone my bad. I shared the wrong error log. I've updated the question. There seems to be no spelling mistake.

To your second comment, yeah I got your point. Since `DeployRaffle.s.sol ` will deploy `AddConsumer.s.sol` contract post `Raffel.sol` it will return the address of contract `AddConsumer` instead of `Raffle's` address. I shall modify my test, but I need to solve this issue of foundry not identifying my `testIfRaffleDeployed()` test.

**Reply by theirrationalone** [reply]
I answered above. But once again... 

Please rename this...   `testRaffleDeployed` ❌ to 

`testIfRaffleDeployed` ☑

You're missing `If` with capital `I`.

**Reply by alfheimrShiven** [reply]
@theirrationalone where am I missing `If`, could you please point it out?
Nowhere there is `testRaffleDeployed` in my updated question. Requesting you to revisit.

**Reply by theirrationalone** [reply]
Oooh Sorry my bad 😅

**Reply by alfheimrShiven** [reply]
Haha no issues.😅

**Reply by shikhar229169** [reply]
Is the file saved?
Idk you have autosaver or not.

**Reply by alfheimrShiven** [reply]
Yes @shikhar229169

**Reply by shikhar229169** [reply]
Can you push your code to github, it's difficult to figure out without having the code.

**Reply by alfheimrShiven** [reply]
Yes @shikhar229169 it's on my Github. You've found it by now🙌🏾

**Comment by mibiot** [comment]
by the way --mt works as well. I use this short command for testing

**Comment by theirrationalone** [comment]
You can use `forge test` only, to check if all tests foundry is taking in account Or not. 

I think that `--match-test` maybe not working you can also try `--mt` for the sake of hit & trial 😅

Thus as far, ever I tried its short version `--mt` and that worked for me each and everytime that's why i can't say anything about `--match-test`.

**Reply by alfheimrShiven** [reply]
No luck. `forge test` run is also not picking up the test.

**Reply by theirrationalone** [reply]
Try to update wsl(if using) and foundry too... 

Wsl: 
```terminal
sudo apt update
```

foundry(you knew it):
```terminal
foundryup
```

**Comment by theirrationalone** [comment]
It would be silly suggestion 😅 but you can try it if you're agree... 


Maybe foundry only reads/catches Folders with UpperCased names.

that `integration` Folder with lowercased `i` in test directory. 

Such a silly suggestion but if works then, I be like 😎 **tHe DoN**

**Reply by shikhar229169** [reply]
So, I actually got the repo from the profile, and switched to commit which contains the DeployTest, and surprisingly it is working on my system, the test failed but the main thing the function was picked up in my system.
Maybe you are not in the required directory. Check that in the command line.

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/111875681/ac061822-947e-45a7-ae8c-c3ee26a2fa88)

**Reply by theirrationalone** [reply]
I checked out his repo, He hasn't committed & Pushed that file on to the Github. Woow He is testing uncommitted file (Bug/Omission). Great Catch @shikhar229169 🚀🚀

**Reply by theirrationalone** [reply]
That's why, I often use `forge init --no-commit`.

**Reply by alfheimrShiven** [reply]
Hey @shikhar229169 lucky you😅. Thanks for the effort man, but I checked out my previous commit which has the `testIfRaffleDeployed()` test and ran `forge test --mt testIfRaffleDeployed` but no luck. I'm sharing my logs below for your reference.

```zsh
shiven@Shivendras-MacBook-Pro lottery-smart-contract % forge test --mt testIfRaffleDeployed
[⠒] Compiling...
[⠆] Compiling 1 files with 0.8.20
[⠰] Solc 0.8.20 finished in 1.66s
Compiler run successful with warnings:
Warning (5667): Unused function parameter. Remove or comment out the variable name to silence this warning.
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:283:47:
    |
283 |     function requestSubscriptionOwnerTransfer(uint64 _subId, address _newOwner) external pure override {
    |                                               ^^^^^^^^^^^^^

Warning (5667): Unused function parameter. Remove or comment out the variable name to silence this warning.
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:283:62:
    |
283 |     function requestSubscriptionOwnerTransfer(uint64 _subId, address _newOwner) external pure override {
    |                                                              ^^^^^^^^^^^^^^^^^

Warning (5667): Unused function parameter. Remove or comment out the variable name to silence this warning.
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:287:46:
    |
287 |     function acceptSubscriptionOwnerTransfer(uint64 _subId) external pure override {
    |                                              ^^^^^^^^^^^^^

Warning (5667): Unused function parameter. Remove or comment out the variable name to silence this warning.
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:291:35:
    |
291 |     function pendingRequestExists(uint64 subId) public view override returns (bool) {
    |                                   ^^^^^^^^^^^^

Warning (2018): Function state mutability can be restricted to pure
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:238:5:
    |
238 |     function getConfig()
    |     ^ (Relevant source part starts here and spans across multiple lines).

Warning (2018): Function state mutability can be restricted to pure
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:251:5:
    |
251 |     function getFeeConfig()
    |     ^ (Relevant source part starts here and spans across multiple lines).

Warning (2018): Function state mutability can be restricted to pure
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:279:5:
    |
279 |     function getFallbackWeiPerUnitLink() external view returns (int256) {
    |     ^ (Relevant source part starts here and spans across multiple lines).

Warning (2018): Function state mutability can be restricted to pure
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:291:5:
    |
291 |     function pendingRequestExists(uint64 subId) public view override returns (bool) {
    |     ^ (Relevant source part starts here and spans across multiple lines).


No tests match the provided pattern:
        match-test: `testIfRaffleDeployed`

Did you mean `testRaffleRecordsPlayerWhenTheyEnter`?
shiven@Shivendras-MacBook-Pro lottery-smart-contract % 
```

**Comment by alfheimrShiven** [comment]
> I checked out his repo, He hasn't committed & Pushed that file on to the Github. Woow He is testing uncommitted file (Bug/Omission). Great Catch @shikhar229169 🚀🚀

@theirrationalone that's not the case. You're checking my latest commit in which I purposefully removed the `DeployRaffleTest.t.sol` as it was not being picked up foundry anyway. Do read the commit msgs to understand.

1. You can checkout to this commit which has the test file:
[https://github.com/alfheimrShiven/lottery-smart-contract/commit/3d162438dacbefd07c70a924e849faac2de0f87c](url)

Commit ID: `3d162438dacbefd07c70a924e849faac2de0f87c`

2. You can now try running `forge test --mt testIfRaffleDeployed` and see if it's been found on your system. Thanks!

**Reply by theirrationalone** [reply]
Ohhhhh Goddddd, You're recursively omitting "" <- Double quotes. 😣😣😣that could also be an issue.

```terminal
forge test --match-test "testIfRaffleDeployed"
```

Actually i paid attention just now.

**Reply by theirrationalone** [reply]
but it is working in machine, In either case...

<br>
<br>

![test ss](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/b5a9f68d-fb83-4903-8f2d-6bc70a37911c)

**Reply by alfheimrShiven** [reply]
That's not the issue @theirrationalone . I would suggest that you clone my repo and try out the solution on your machine before suggesting it here. That would keep this thread clean and easier for others to contribute as well.

**Reply by shikhar229169** [reply]
Try these things:

- Run ```forge tree``` and check if that test file is there or not.
- Move that test file from integration to unit folder, and try to run test.

**Comment by theirrationalone** [comment]
Hey i got the Issue...

Actually Foundry treats all files as a script only whose names start with Deploy.

## for example: `DeployXYZ.t.sol` <- files named like this will always treated as a Script file.

try to rename your testing script like this.

![mitigated](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/c564e375-cffa-423c-8f26-675ea7587bf0)

**Reply by theirrationalone** [reply]
And you didn't toggled that ffi in `foundry.toml`

```toml
ffi = true
```

and as i discussed above that `most_recent_deployment` reads contract from that `broadcast` folder.

so as the result...

![failiing](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/cd7cbe2c-fee4-416d-93a5-07bff6cf7f48)

**Reply by alfheimrShiven** [reply]
@theirrationalone thats not necessary here as referred from Pattrick's codebase:
[https://github.com/Cyfrin/foundry-smart-contract-lottery-f23/blob/main/foundry.toml](url)

**Reply by theirrationalone** [reply]
The issue is that prepending file names with word "Deploy" that's why foundry treating that file as a deploy script and not taking that file as a test script in account.

**Comment by Niferu** [comment]
Hi @alfheimrShiven 

I'm not sure whats wrong with your `DeployRaffleTest.t.sol` but it is somehow corrupted.

Solution for your issue:
1. Manually create new file in test->unit folder named `DeployRaffleTest.t.sol`
2. Copy code from your corrupted file into newly created one and run `forge test --mt testIfRaffleDeployed`

In case above wont work, in unit folder create file named `NewTest.t.sol` copy code from your `DeployRaffleTest` and rename test function from `testIfRaffleDeployed` into `testRaffleDeployed`.

Both solutions worked for me, but I admit it is very bizzare..

**Comment by alfheimrShiven** [comment, marked as answer]
Well @shikhar229169 @theirrationalone @Neftyr @mibiot , I identified the issue. The issue was that the contract names for two of my test files were **identical**. 
Therefore, Foundry was not reading one of the test files (`tests/integration/DeployRaffleTest.t.sol`) whose contract name was the same as a previously complied test file(`tests/unit/InteractionsTest.t.sol`).
Thanks for your support. Really appreciated 🙌🏾. Hope we all learned to be careful of not duplicating our test contract names 🤣. 

Attaching a screenshot from that commit for you ref:
<img width="1440" alt="Screenshot 2023-08-09 at 8 50 38 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/50408041/9a61c839-257c-4dea-969a-b5d117ae522e">

**Reply by theirrationalone** [reply]
I had cloned your repo, i hadn't got that issue but got that very same you discussed in this discussion.

And i successfully mitigated that... Look this comment...  [Comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/582#discussioncomment-6678407)

**Reply by alfheimrShiven** [reply]
@theirrationalone if you visit `/test/unit/InteractionsTest.t.sol` in my clone on your system, you will see the identical contract name issue I'm mentioning. 
The reason you gave and I quote _"Actually Foundry treats all files as a script only whose names start with Deploy."_  isn't correct. The issue wasn't due to the name starting with 'Deploy'. 
Hope I've explained it well. Thanks.

**Reply by theirrationalone** [reply]
Ooooh I see It's different from hardhat.woww thanks for clearing this too.

**Reply by Niferu** [reply]
Hmm but that solution is pretty strange as I cloned your repo and I have just created new file named same and with same contract name (so it was duplicated still) and it worked for me I didnt change anything else...

**Reply by theirrationalone** [reply]
Same in my case too. Foundry not raising issue on identical contracts. Maybe they're scoped to file only.

**Reply by shikhar229169** [reply]
+1 Same in my case.

**Reply by alfheimrShiven** [reply]
Hmm. I see. Anyway. It works now. Will think about this further if I encounter such an issue moving forward. Thanks guys!
